internal/jobcontainers/process.go

```diff
@@ -0,0 +1,228 @@
+	return p.stdin.Close()
+}
+
+// Wait waits for the process to exit. If the process has already exited returns
+// the previous error (if any).
+func (p *JobProcess) Wait() error {
+	<-p.waitBlock
+	return p.waitError
+}
+
+// Start starts the job object process
+func (p *JobProcess) Start() error {
+	return p.cmd.Start()
+}
+
+// This should only be called once.
+func (p *JobProcess) waitBackground(ctx context.Context) {
+	log.G(ctx).WithField("pid", p.Pid()).Debug("waitBackground for JobProcess")
+
+	// Wait for process to get signaled/exit/terminate/.
+	err := p.cmd.Wait()
+
+	// Wait closes the stdio pipes so theres no need to later on.
+	p.stdioLock.Lock()
+	p.stdin = nil
+	p.stdout = nil
+	p.stderr = nil
+	p.stdioLock.Unlock()
+
+	p.closedWaitOnce.Do(func() {
+		p.waitError = err
+		close(p.waitBlock)
+	})
+}
+
+// ExitCode returns the exit code of the process.
+func (p *JobProcess) ExitCode() (int, error) {
+	p.procLock.Lock()
+	defer p.procLock.Unlock()
+
+	if !p.exited() {
+		return -1, errors.New("process has not exited")
+	}
+	return p.cmd.ProcessState.ExitCode(), nil
+}
+
+// Pid returns the processes PID
+func (p *JobProcess) Pid() int {
+	if process := p.cmd.Process; process != nil {
+		return process.Pid
+	}
+	return 0
+}
+
+// Close cleans up any state associated with the process but does not kill it.
+func (p *JobProcess) Close() error {
+	p.stdioLock.Lock()
+	if p.stdin != nil {
+		p.stdin.Close()
+		p.stdin = nil
+	}
+	if p.stdout != nil {
+		p.stdout.Close()
+		p.stdout = nil
+	}
+	if p.stderr != nil {
+		p.stderr.Close()
+		p.stderr = nil
+	}
+	p.stdioLock.Unlock()
+
+	p.closedWaitOnce.Do(func() {
+		p.waitError = hcs.ErrAlreadyClosed
+		close(p.waitBlock)
+	})
+	return nil
+}
+
+// Kill kills the running process. Go calls TerminateProcess under the hood.
+func (p *JobProcess) Kill(ctx context.Context) (bool, error) {
+	log.G(ctx).WithField("pid", p.Pid()).Debug("killing job process")
+
+	p.procLock.Lock()
+	defer p.procLock.Unlock()
+
+	if p.exited() {
+		return false, errors.New("kill not sent. process already exited")
+	}
+
+	if p.cmd.Process != nil {
+		if err := p.cmd.Process.Kill(); err != nil {
+			return false, err
+		}
+	}
+	return true, nil
+}
+
+func (p *JobProcess) exited() bool {
+	if p.cmd.ProcessState == nil {
+		return false
+	}
+	return p.cmd.ProcessState.Exited()
+}
+
+// signalProcess sends the specified signal to a process. We can't use GenerateConsoleCtrlEvent since
+// that only supports CTRL_C_EVENT and CTRL_BREAK_EVENT. Instead to handle an arbitrary signal,
+// we open a CtrlRoutine thread inside the target process and give it the specified signal to handle.
+//
+// Note: This is a hack which is not officially supported.
+func signalProcess(pid uint32, signal int) error {
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, pid)
+	if err != nil {
+		return errors.Wrap(err, "failed to open process")
+	}
+	defer windows.Close(hProc)
+
+	k32, err := windows.LoadLibrary("kernel32.dll")
```

Comment:
Is there any consideration for using kernel32 rather than kernelbase?

Reply:
It looks like kernel32 exports CtrlRoutine as a forwarder to kernelbase.

Reply:
Ping

Reply:
No consideration beyond the console folks said either would work/either would be a fine approach.

Reply:
Sorry, this got buried